### PR TITLE
[WIP] Adds service workers to Calendars App

### PR DESCRIPTION
**Pull request by DilwoarH**
This allows `/bank-holidays` and `/when-do-the-clocks-change` pages to be cached offline
so that users hitting this page do not keep making requests to servers and have
offline capabilities.

Screen Shot 2019-04-30 at 12 43 29

This work was done as part of ongoing frontend community improvements to GOV.UK.

https://trello.com/c/ZGRXpx6P/87-implementing-experimenting-with-service-workers-on-govuk

## Demo
https://govuk-calendars-pr-560.herokuapp.com/bank-holidays

---

Credits: https://deanhume.com/create-a-really-really-simple-offline-page-using-service-workers/

### CODE REVIEW

**Comment by Nooshu** [comment]
Nice work @DilwoarH.

We should definitely include a `activate` event that will clean up redundant caches. In the article linked to it gives a working version that can be seen [here](https://github.com/GoogleChrome/samples/blob/gh-pages/service-worker/custom-offline-page/service-worker.js#L51)

We should also make sure that the `service-worker.js` file isn't bundled with any other JavaScript and has `Cache-Control: no-cache` and `Cache-Control: max-age=0` headers set. This is actually part of the SW spec, but always good to have for "older" evergreen browsers (users locked to a particular version).

**Comment by andysellick** [comment]
@DilwoarH do you want a review already? Doesn't seem like this is ready yet :)

**Comment by tombye** [comment]
As a discussion point, I’d be in favour of adding a kill-switch.

Jeff Posnick put up some comments about this: https://stackoverflow.com/questions/33986976/how-can-i-remove-a-buggy-service-worker-or-implement-a-kill-switch/38980776#38980776

I know this specific code is pretty harmless but I think there's some value in starting to look at what our answer would be to a scenario where service worker JS is failing and we need to purge it from all clients.

**Comment by Nooshu** [comment]
@tombye could be as simple as having a `debug` variable with the no-op code mentioned in the linked SO answer that would:

- Force skip waiting on install with no other functionality
- Force reload all current open browser tabs (listed as optional in the article)
- Purge the SW Cache for the given domain on activation (could look something like this):

```js
addEventListener('activate', activateEvent => {
    activateEvent.waitUntil(
        caches.keys()
        .then(cacheNames => {
            return Promise.all(
                cacheNames.map( cacheName => {
                    return caches.delete(cacheName);
                })
            ).then(response => {
                // do something else here if required
            })
        })
    );
});
```
As mentioned in the article all current evergreen browsers no-longer use the HTTP cache for the SW, and will always return to the network for an updated version. If there's a single byte difference in the SW it is seen as a new version and will nuke the old version and install the new.

File: public/service-worker.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
+// https://deanhume.com/create-a-really-really-simple-offline-page-using-service-workers/
+'use strict';
+
+var cacheVersion = 1;
+var currentCache = {
```
Comment (ChrisBAshton) [inline public/service-worker.js]: Any reason these are `var` when we're using `const` on line 8?

File: public/service-worker.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  );
+});
+
+this.addEventListener('fetch', event => {
+  // request.mode = navigate isn't supported in all browsers
+  // so include a check for Accept: text/html header.
+  if (event.request.mode === 'navigate' || (event.request.method === 'GET' && event.request.headers.get('accept').includes('text/html'))) {
+    event.respondWith(
+      fetch(event.request.url).catch(error => {
+        // Return the offline page
```
Comment (ChrisBAshton) [inline public/service-worker.js]: This reads like we're serving a "You are offline, please hit 'back'" kind of page, when really we're serving the copy of the page the user has visited in the past, if I've understood this...? 😄 

```suggestion
        // Return the cached page if it exists
```

File: public/service-worker.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        return caches.match(event.request.url);
+      })
+    );
+  } else {
+    // Respond with everything else if we can
```
Comment (ChrisBAshton) [inline public/service-worker.js]: Had to re-read this to understand that this isn't returning pages, but assets.

```suggestion
    // This is an asset - attempt to serve from cache, fall back to network
```

Love this PR! Can't wait to see it merged 🎉

File: public/service-worker.js (hunk 1 above)
Comment (Nooshu) [reply public/service-worker.js]: Should be safe to use all ES2015+ features in the service worker file (assuming we wrap appropriately using something like native JavaScript modules). All major browsers that support service workers also support native JavaScript modules, so it's perfectly fine to serve this code to any browsers (older browsers will just ignore it).

Comment (ChrisBAshton) [reply public/service-worker.js]: My point is more that these should all be `const`, rather than a mixture of `var` and `const`.

Comment (Nooshu) [reply public/service-worker.js]: I agree

**Comment by DilwoarH** [comment]
From Matt:
https://offline-documentation-passive.netlify.com/service-worker.js

File: public/service-worker.js (hunk 1 above)
Comment (DilwoarH) [reply public/service-worker.js]: Turns out we can't use `const` in GOV.UK:
```
[docker] Uglifier::Error: Unexpected token: keyword (const). To use ES6 syntax, harmony mode must be enabled with Uglifier.new(:harmony => true).
```

**Comment by DilwoarH** [comment]
@Nooshu @tombye Do you know if we have to add a service worker policy (like the cookies policy) on GOV.UK?

**Comment by Nooshu** [comment]
@DilwoarH There's no obligation to add anything. No personal data is being stored. When it comes to caching of assets it's really no different from storing them using standard cache control headers (we just have fine grain control over it). Either way the asset is being stored on a users machine.

I searched to see if there's an example of others mentioning the use of a SW on their website, but couldn't find anything.

**Comment by nickcolley** [comment]
This seems very complicated and could cause issues if we've made a mistake with the caching.

Do we have clear user needs for this work? Do people need to come back to this page frequently for this feature to be worth the overhead?

**Comment by DilwoarH** [comment]
Closing for now as this work needs a lot of GOV.UK architectural changes

**Pull request closed without merging**